README
SVG-Rasterize

Rasterize SVG content to pixel graphics.

SVG::Rasterize can be used to rasterize SVG objects to pixel
graphics building on the perl bindings of the cairo library (by
default, other underlying rasterization engines could be added). The
direct rasterization of SVG _files_ might be implemented in the
future, right now you should have a look at SVG::Parser which can
generate an SVG object from an svg file.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

CHANGES

For a list of changes have a look at the separate Changes file.

DEPENDENCIES

SVG::Rasterize depends on the following modules:

Class::Accessor,  version 0.30  or higher
SVG,              version 2.37  or higher
Cairo,            version 1.061 or higher
Pango,            version 1.220 or higher
Params::Validate, version 0.91  or higher
Scalar::Util,     version 1.19  or higher
Exception::Class, version 1.29  or higher
Test::More,       version 0.86  or higher
Test::Exception   version 0.27  or higher
Test::Warn        version 0.08  or higher

Furthermore, the underlying cairo and pango C libraries are required
at least in version 1.8.8 and version 1.22.4, respectively. These
dependencies are not automatically fulfilled by installing a
sufficiently high version of the respective Perl module. For pango,
the functionality of version 1.22 is indeed required. In contrast,
cairo version 1.2 might actually be enough. However, 1.8 was the
smallest version I got pango compiled with.

With respect to the module code, the dependency on Cairo and Pango
is not strict. Both are loaded by SVG::Rasterize::Cairo and that
module is only loaded dynamically if no other rasterization backend
has been specified (see documentation for details). However, if you
do not provide a different backend, which would probably at least
require a wrapper written by you, then you cannot do anything
without SVG::Rasterize::Cairo (and thereby Cairo and
Pango). Therefore I have included them as a strict dependency. You
could take them out of the Makefile.PL if you know what you are
doing. However, the distribution will not pass the test suite
without them.

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with
the perldoc command.

    perldoc SVG::Rasterize

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=SVG-Rasterize

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/SVG-Rasterize

    CPAN Ratings
        http://cpanratings.perl.org/d/SVG-Rasterize

    Search CPAN
        http://search.cpan.org/dist/SVG-Rasterize/


Copyright 2010 Lutz Gehlen.

This program is free software; you can redistribute it and/or modify
it under the terms of either: the GNU General Public License as
published by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for SVG-Rasterize

0.003007 Tue May  3 19:51:07 NZST 2011
	 - Incompatible changes:
	   - The default backend SVG::Rasterize::Cairo has been
	     renamed to SVG::Rasterize::Engine::PangoCairo. There is
	     a base class SVG::Rasterize::Engine which backends
	     should inherit from. If you are selecting the
	     rasterization backend explicitly you will have to
	     update your code. Otherwise - unless you are
	     maintaining your own rasterization backend for
	     SVG::Rasterize - this change is completely transparent
	     for you.
	 - more font attributes
	 - minor fixes and improvements
	 - docu expansion

0.003006 Thu Apr 28 20:21:15 NZST 2011
	 - Incompatible changes:
	   - The current_text_position method of
	     SVG::Rasterize::State has gone.
	 - Fixes bug #67699.
	 - Major revision of the traversal of the node tree enabling
	   a much cleaner approach to deferred rasterization of text
	   elements. This should be transparent even for hooks.
	 - Major revision of the way text is processed. This
	   framework will hopefully be fully capable of handling
	   text correctly including bidi text etc..
	 - Alignment of text via text-anchor
	 - Positioning of individual characters

0.003005 Fri Apr 22 11:53:35 NZST 2011
	 - Support for more text attributes:
	   - font-size (except relative font sizes)
	 - Regex check of ID attributes.
	 - Reorganized docu to increase user friendliness.
	 - Reorganization and substantial expansion of test suite,
	   guided by Devel::Cover.
	 - Comments are ignored now as they should. This fixes bug
	   RT#62943. Additionally, title, desc, and metadata
	   elements are ignored.
	 - Minor bug fixes.

0.003004 Thu Jun 17 13:16:14 NZST 2010
	 - Adds support for currentColor and hex color settings.
	 - Introduces methods to restore default hooks.
	 - Makefile.PL tests for sufficiently high versions of the
	   cairo and pango C libraries thereby avoiding failing
	   CPAN testers smoke tests.

0.003003 Sun Jun 13 20:48:11 NZST 2010
	 This is a maintenance release covering the following
	 improvements:
	 - The specification is split into different modules
	   (according to the classes in the DTD) which are only
	   loaded when needed. The avoids loading the 12KB
	   Specification module at the start even if only small bits
	   will be used.
	 - The specification modules now contain information which
	   elements are allowed to hold parsed character data (other
	   than white space). This is checked during validation.
	 - Class::Accessor now also throws an Exception::Class
	   object if it is attempted to set a readonly attribute.

0.003002 Fri Jun 11 20:09:34 NZST 2010
	 - Now introducing the Pango dependency (see 0.003001).
	 - Starting support for text elements. This brings some
	   little, but not backwards compatible changes in the
	   interface of SVG::Rasterize::State:
	   - The child nodes are determined outside of state now and
             handed over to the constructor.
	   - The method hasChildNodes has gone, and nextChild has
             been renamed to shift_child_node.
	 - Additionally, there has been an interface change of the
           before_node_hook (which was marked as experimental). It
	   now receives named parameters and is expected to return a
	   (potentially modified) hash.
	 - Various bug fixes including RT#58114.
	 - Detailed audit of the main docu.

0.003001 Thu Jun  3 20:36:38 NZST 2010
	 - Main purpose for this (kind of premature) release: Fixes
	   bug in SVG::Rasterize::Cairo that prohibit the
	   specification of an output file name.
	 - New dependency Pango (1.220 or higher)...postponed in
	   order to allow testing my perl 5.8.9 installation.
	 - Fixes bug in validation of attributes with a list of
	   valid values.

0.003000 Wed Jun  2 20:41:08 NZST 2010
	 Identical to 0.002003.

0.002003 Wed Jun  2 20:34:37 NZST 2010 (unpublished)
	 - Prepared the framework for greater flexibility in output
           formats etc..
	 - Docu fixes and expansion.

0.002002 Tue Jun  1 21:08:09 NZST 2010 (unpublished)
	 - Partial rendering of paths, polylines, and polygons if
	   they are in error.
	 - Consequent use of Params::Validate (or documentation that
	   parameters are not validated).
	 - Documentation of diagnostics.
	 - Docu fixes and expansion.

0.002001 Sun May 30 16:21:15 NZST 2010 (unpublished)
	 - New dependency Exception::Class (1.29 or higher).
	 - Proper error handling as required by spec (except partial
	   rendering of path, polyline, and polygon).
	 - Docu fixes.
	 - All methods return explicitly.

0.002000 Tue May 25 15:42:09 NZST 2010
	 Documentation updates.

0.001006 Tue May 25 15:06:16 NZST 2010 (unpublished)
	 - Adds support for all basic shapes
	   - rect
	   - circle
	   - ellipse
	   - (line was already supported)
	   - polyline
	   - polygon
	 - Revision of large parts of the docu.

0.001005 Sun May 23 21:39:39 NZST 2010
	 - Color filling (no gradients, patterns etc.).
	 - (Semi-)transparent stroking and filling including
	   different opacity values for stroke and fill and without
	   the fill shining through transparent stroke.
	 - Further newly supported properties (all fill and stroke
	   properties):
	   - fill-rule
	   - stroke-linecap
	   - stroke-linejoin
	   - stroke-miterlimit
	   - stroke-dashoffset
	   - stroke-dasharray
	 - Bugfixes:
	   - in processing of explicit 'inherit' property
	     values
	   - in path data processing
	   - color values are clamped at presentation time instead
	     of parsing time (as recommended by the spec).

0.001004 Fri May 21 21:19:54 NZST 2010 (unpublished)
	 - Elliptical arcs including example script.

0.001003 Thu May 20 19:59:30 NZST 2010 (unpublished)
	 - Support for all path commands except elliptical arc.
	 - Core documentation up-to-date.

0.001002 Sun May 16 19:33:22 NZST 2010
	 - The final transformation matrix is now applied by the
	   rasterization engine and not by SVG::Rasterize. This is

Makefile.PL
use inc::Module::Install;

name     'SVG-Rasterize';
all_from 'lib/SVG/Rasterize.pm';
author   q{Lutz Gehlen <perl@lutzgehlen.de>};
license  'perl';

no_index directory => 'examples';

requires           'Class::Accessor'  => '0.30';
configure_requires 'Cairo'            => '1.061';
configure_requires 'Pango'            => '1.220';
requires           'Params::Validate' => '0.91';
requires           'Scalar::Util'     => '1.19';
requires           'Exception::Class' => '1.29';

test_requires      'SVG'              => '2.37';
test_requires      'Test::More'       => '0.80';
test_requires      'Test::Exception'  => '0.27';

# for systems not supporting configure_requires we require Pango
# within an eval statement
if(eval { require Cairo }) {
    if(Cairo->VERSION < Cairo->VERSION_ENCODE (1, 8, 8)) {
	print("SVG::Rasterize requires the Cairo C library at least ",
	      "in version 1.8.8. Please upgrade.\n");
	print(STDERR "NA: Unable to build distribution on this ".
	      "platform.\n");
	exit(0);
    }
}
else {
    print("The execution of the SVG::Rasterize Makefile.PL requires ", 
	  "the presence of the Cairo module. Your system does not seem ",
	  "to support configure_requires entries in META.YML. Please ".
	  "install Cairo before installing SVG::Rasterize.\n");
    print(STDERR "NA: Unable to build distribution on this ".
	  "platform.\n");
    exit(0);
}

# for systems not supporting configure_requires we require Pango
# within an eval statement
if(eval { require Pango }) {
    if(!Pango->CHECK_VERSION(1, 22, 4)) {
	print("SVG::Rasterize requires the Pango C library at least ",
	      "in version 1.22.4. Please upgrade.\n");
	print(STDERR "NA: Unable to build distribution on this ".
	      "platform.\n");
	exit(0);
    }
}
else {
    print("The execution of the SVG::Rasterize Makefile.PL requires ", 
	  "the presence of the Pango module. Your system does not seem ",
	  "to support configure_requires entries in META.YML. Please ".
	  "install Pango before installing SVG::Rasterize.\n");
    print(STDERR "NA: Unable to build distribution on this ".
	  "platform.\n");
    exit(0);
}

WriteAll;
